File: frontend/src/app/pages/employees/employees.ts
import { Component, OnInit, ChangeDetectorRef } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';

import { EmployeeService, Employee } from '../../services/employee.service';

@Component({
  selector: 'app-employees',
  standalone: true,
  imports: [CommonModule, FormsModule],
  templateUrl: './employees.html',
  styleUrl: './employees.css'
})
export class Employees implements OnInit {

  employees: Employee[] = [];

  searchText = '';

  isEditing = false;

  newEmployee: Employee = {
    full_name: '',
    employee_id: '',
    email: '',
    phone: '',
    department: '',
    role: '',
    joining_date: '',
    status: 'Active'
  };

  constructor(
    private employeeService: EmployeeService,
    private cdr: ChangeDetectorRef
  ) {}

  ngOnInit(): void {
    this.loadEmployees();
  }
loadEmployees(): void {

  console.log("🚀 loadEmployees() called");

  this.employeeService.getEmployees().subscribe({

    next: (data) => {

      console.log("✅ API Response:", data);

      this.employees = data;

      this.cdr.detectChanges();

    },

    error: (err) => {

      console.error("❌ Error:", err);

    }

  });

}
  get filteredEmployees(): Employee[] {

    if (!this.searchText.trim()) {
      return this.employees;
    }

    const search = this.searchText.toLowerCase();

    return this.employees.filter(emp =>

      emp.full_name.toLowerCase().includes(search) ||

      emp.employee_id.toLowerCase().includes(search) ||

      emp.department.toLowerCase().includes(search) ||

      emp.role.toLowerCase().includes(search)

    );

  }

  get activeEmployees(): number {
    return this.employees.filter(e => e.status === 'Active').length;
  }

  get leaveEmployees(): number {
    return this.employees.filter(e => e.status === 'Leave').length;
  }

  get departmentCount(): number {

    return new Set(
      this.employees.map(emp => emp.department)
    ).size;

  }

  addEmployee(): void {

    if (this.isEditing) {

      this.employeeService.updateEmployee(
        this.newEmployee.employee_id,
        this.newEmployee
      ).subscribe({

        next: () => {

          alert("Employee updated successfully");

          this.resetForm();

          this.loadEmployees();

        },

        error: err => console.error(err)

      });

      return;
    }

    this.employeeService.addEmployee(this.newEmployee).subscribe({

      next: () => {

        alert("Employee added successfully");

        this.resetForm();

        this.loadEmployees();

      },

      error: err => console.error(err)

    });

  }

  editEmployee(emp: Employee): void {

    this.newEmployee = { ...emp };

    this.isEditing = true;

  }

  deleteEmployee(employeeId: string): void {

    if (!confirm("Delete this employee?")) {
      return;
    }

    this.employeeService.deleteEmployee(employeeId).subscribe({

      next: () => {

        alert("Employee deleted successfully");

        this.loadEmployees();

      },

      error: err => console.error(err)

    });

  }

  resetForm(): void {

    this.newEmployee = {
      full_name: '',
      employee_id: '',
      email: '',
      phone: '',
      department: '',
      role: '',
      joining_date: '',
      status: 'Active'
    };

    this.isEditing = false;

  }

}

File: app/api/routes/employees.py
from fastapi import APIRouter, HTTPException

from app.database.employee import (
    save_employee,
    get_all_employees,
    get_employee_details,
    delete_employee
)

router = APIRouter(
    prefix="/employees",
    tags=["Employees"]
)


@router.get("/")
def list_employees():

    employees = get_all_employees()

    return [
        dict(employee)
        for employee in employees
    ]


@router.post("/")
def add_employee(employee: dict):

    save_employee(
        employee["full_name"],
        employee["employee_id"],
        employee["email"],
        employee["phone"],
        employee["department"],
        employee["role"],
        employee["joining_date"],
        employee["status"]
    )

    return {
        "message": "Employee added successfully"
    }


@router.get("/{employee_id}")
def get_employee(employee_id: str):

    employee = get_employee_details(employee_id)

    if employee is None:
        raise HTTPException(
            status_code=404,
            detail="Employee not found"
        )

    return dict(employee)


@router.delete("/{employee_id}")
def remove_employee(employee_id: str):

    employee = get_employee_details(employee_id)

    if employee is None:
        raise HTTPException(
            status_code=404,
            detail="Employee not found"
        )

    delete_employee(employee_id)

    return {
        "message": "Employee deleted successfully"
    }